### PR TITLE
Refactor and rewrite clasp run and manifest tests

### PR DESCRIPTION
**Pull request by campionfellin**
Refactors the `clasp run` tests, as well as a number of functions within `run.ts` and `manifest.ts`. Comments inline.

Increases overall coverage from:

```
-----------------|----------|----------|----------|----------|-------------------|
File             |  % Stmts | % Branch |  % Funcs |  % Lines | Uncovered Line #s |
-----------------|----------|----------|----------|----------|-------------------|
All files        |    73.85 |    49.83 |    77.17 |    75.66 |                   |
 src             |    67.84 |    54.69 |    62.69 |    70.31 |                   |
```
to
```
-----------------|----------|----------|----------|----------|-------------------|
File             |  % Stmts | % Branch |  % Funcs |  % Lines | Uncovered Line #s |
-----------------|----------|----------|----------|----------|-------------------|
All files        |    76.69 |    52.55 |    81.22 |    78.38 |                   |
 src             |    71.22 |    59.23 |    68.61 |    73.78 |                   |
```

Specifically, `clasp run` coverage from:

`  run.ts         |    12.33 |        0 |        0 |    13.11 |`

to

`run.ts         |    37.68 |    11.54 |       50 |    35.09 |`

and manifest from:

`manifest.ts    |    62.96 |    54.17 |    66.67 |    65.08 |`

to

`manifest.ts    |    67.05 |    60.71 |    72.73 |    69.12 |`

and utils from:

`utils.ts       |    66.67 |    53.93 |    61.67 |    65.71 |`

to

`utils.ts       |     67.5 |    56.18 |     62.3 |    66.67 |`

Signed-off-by: campionfellin <campionfellin@gmail.com>

- [x] `npm run test` succeeds.
- [x] `npm run lint` succeeds.
- [ ] Appropriate changes to README are included in PR.

### CODE REVIEW

File: src/commands/run.ts [hunk 1]
```diff
@@ -10,46 +10,35 @@ import {
...
 /**
  * Executes an Apps Script function. Requires clasp login --creds.
  * @param functionName {string} The function name within the Apps Script project.
  * @param cmd.nondev {boolean} If we want to run the last deployed version vs the latest code.
+ * @param cmd.params {string} JSON string of parameters to be input to function.
  * @see https://developers.google.com/apps-script/api/how-tos/execute
  * @requires `clasp login --creds` to be run beforehand.
  */
 export default async (functionName: string, cmd: { nondev: boolean; params: string }) => {
-  function IsValidJSONString(str: string) {
```
Comment (campionfellin) [inline src/commands/run.ts]: Moved this to `src/utils.ts` and changed it to `getValidJSON()`.

File: src/commands/run.ts [hunk 2]
```diff
@@ -10,46 +10,35 @@ import {
...
-  const params = JSON.parse(paramString);
-  // Ensures the manifest is correct for running a function.
-  // The manifest must include:
-  // "executionApi": {
-  //   "access": "MYSELF"
-  // }
-  await isValidManifest();
+  const params = getValidJSON(paramString);
+
+  await isValidRunManifest();
```
Comment (campionfellin) [inline src/commands/run.ts]: `isValidManifest` function simply tested that it was valid JSON. This makes sure it includes 
```
"executionApi": {	
    "access": "MYSELF"	
  }
```

as noted in the comment.

File: src/commands/run.ts [hunk 3]
```diff
@@ -59,113 +48,115 @@ export default async (functionName: string, cmd: { nondev: boolean; params: stri
...
-        }
-      } else if (data.error && data.error.details) {
-        // @see https://developers.google.com/apps-script/api/reference/rest/v1/scripts/run#Status
-        console.error(
-          `${chalk.red('Exception:')}`,
-          data.error.details[0].errorType,
-          data.error.details[0].errorMessage,
-          data.error.details[0].scriptStackTraceElements || [],
-        );
+  await runFunction(functionName, params, scriptId, devMode);
```
Comment (campionfellin) [inline src/commands/run.ts]: Moved `runFunction` out of the main function into its own. No changes here.

File: src/manifest.ts [hunk 4]
```diff
@@ -46,22 +47,40 @@ export async function writeManifest(manifest: Manifest) {
...
+
+/**
+ * Reads manifest file from project root dir.
  * The manifest is valid if it:
  * - It exists in the project root.
  * - Is valid JSON.
  */
-export async function isValidManifest(): Promise<boolean> {
+// tslint:disable-next-line:no-any
+export async function getManifest(): Promise<any> {
```
Comment (campionfellin) [inline src/manifest.ts]: `isValidManifest` previously just checked if it was a valid JSON file, which is what `getValidJSON` does anyway. This returns a manifest if its valid.

File: src/commands/run.ts (hunk 2 above)
Comment (erickoledadevrel) [reply src/commands/run.ts]: Should the other access levels be allowed to? If you've deployed it to run for your entire domain, then that should be good enough for clasp run.

File: src/manifest.ts [hunk 5]
```diff
@@ -46,22 +47,40 @@ export async function writeManifest(manifest: Manifest) {
...
+
+/**
+ * Ensures the manifest is correct for running a function.
+ * The manifest must include:
+ * "executionApi": {
+ *   "access": "MYSELF"
+ * }
+ */
+export async function isValidRunManifest(): Promise<boolean> {
+  if (isValidManifest) {
```
Comment (erickoledadevrel) [inline src/manifest.ts]: Shouldn't you be running this function? Doesn't this current like just check if the function is defined?

File: src/commands/run.ts (hunk 2 above)
Comment (campionfellin) [reply src/commands/run.ts]: Definitely. Is there documentation on the other access levels? Or should It just be a blanket of anything containing `executionApi` and `access`, regardless of value of `access`?

File: src/manifest.ts (hunk 5 above)
Comment (campionfellin) [reply src/manifest.ts]: Yes, great catch, should be `if (isValidManifest()) {`. I'll look at adding another test for this as well.

Comment (erickoledadevrel) [reply src/manifest.ts]: And that function returns a promise, so I think you'll also need to await the promise to get the actual boolean result.

File: src/commands/run.ts (hunk 2 above)
Comment (erickoledadevrel) [reply src/commands/run.ts]: The values are documented here:

https://developers.google.com/apps-script/concepts/manifests

All of the values would seem to work for the `clasp run` case, so I don't think we need to check which access is set.

File: tests/constants.ts [hunk 6]
```diff
@@ -9,12 +9,27 @@ import { OAuth2ClientOptions } from 'google-auth-library';
...
+const APPSSCRIPT_JSON_WITH_RUN = JSON.stringify({
+  timeZone: 'America/Los_Angeles',
+  dependencies: {},
+  exceptionLogging: 'STACKDRIVER',
+  executionApi: {
+    access: 'MYSELF',
+  },
+});
+
+export const TEST_APPSSCRIPT_JSON = {
```
Comment (grant) [inline tests/constants.ts]: Please remove this const and export the other constants.

**Pull request merged**